google/cloud/spanner/internal/connection_impl.cc

```diff
@@ -54,15 +71,83 @@ StatusOr<ConnectionImpl::SessionHolder> ConnectionImpl::GetSession() {
   return SessionHolder(std::move(*response->mutable_name()), this);
 }
 
+StatusOr<ResultSet> ConnectionImpl::Read(spanner_proto::TransactionSelector& s,
+                                         ReadParams rp) {
+  auto session = GetSession();
+  if (!session) {
+    return std::move(session).status();
+  }
+  spanner_proto::ReadRequest request;
+  request.set_session(session->session_name());
+  *request.mutable_transaction() = s;
+  request.set_table(std::move(rp.table));
+  request.set_index(std::move(rp.read_options.index_name));
+  for (auto&& column : rp.columns) {
+    request.add_columns(std::move(column));
+  }
+  *request.mutable_key_set() = internal::ToProto(std::move(rp.keys));
+  request.set_limit(rp.read_options.limit);
+
+  grpc::ClientContext context;
+  auto reader = internal::PartialResultSetReader::Create(
+      stub_->StreamingRead(context, request));
+  if (!reader.ok()) {
+    return std::move(reader).status();
+  }
+  if (s.has_begin()) {
+    auto metadata = (*reader)->Metadata();
+    if (!metadata || metadata->transaction().id().empty()) {
+      return Status(StatusCode::kInternal,
+                    "begin transaction requested but no transaction returned");
```

Comment:
Do you think we should include the operation (maybe `__func__` or `Client::Read`) in the error message?